### DISCUSSION TITLE
UDP server: why the main loop doesn't block the program?

### DISCUSSION THREAD

**Discussion by cristian-spiescu** [micropython/micropython#15427] [ESP32] [answered]
Hi guys,

Sorry for this probably dumb question. I'm using an M5 Stack ESP32 board, and I created a quick "hello world" style program that listens for data on UDP. Here it is:

```python
from m5stack import *
from m5ui import *
from uiflow import *
import socket

setScreenColor(0x111111)

label0 = M5TextBox(13, 9, "label0", lcd.FONT_Default, 0xFFFFFF, rotate=0)

def buttonA_wasPressed():
  # global params
  label0.setText('Hello M5')
  pass
btnA.wasPressed(buttonA_wasPressed)

udpsocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
udpsocket.bind(('0.0.0.0', 5000))
while True:
  print(udpsocket.recv(1024))
  wait_ms(2)
```

I'm using the "Ui Flow" web IDE of M5 Stack which has goodies such as a WYSIWYG editor for the screen of the M5 Stack device. Hence `label0`.

**My question:** my understanding is that the `udpsocket.recv()` is a blocking call. Hence my program is blocked in the loop until data exists. However, if I press on the button of the device => `buttonA_wasPressed()` is invoked successfully. And I don't understand why this works. It is like there were 2 threads: one for the loop, and another one for the handler function for the button. But AFAIK the micropython program is single threaded.

Any hints?

Thanks 🙂

**Comment by davefes** [comment]
I would put into:
```py
while True:
  print(udpsocket.recv(1024))
  wait_ms(2)
```
`print('looping')` and change the `wait_ms(2)` to `wait_ms(1000)`
to verify that you are really in that loop.

**Reply by davefes** [reply]
Did a search and I read a few about UDP calls being non-blocking.  Maybe, that doesn't make any difference in your case.
https://github.com/micropython/micropython/issues/5759

**Comment by cristian-spiescu** [comment]
I performed the modifications. I confirm that the line `udpsocket.recv()` is blocked. When data arrives via UDP => a new iteration is performed.

**Reply by davefes** [reply]
Which pin is `buttonA` connected to?  A link to the board schematic could be helpful.

**Reply by cristian-spiescu** [reply]
Looking [here](https://docs.m5stack.com/en/core/m5stickc_plus): GPIO37.

**Reply by davefes** [reply]
Thanks, I see I can't "translate" that to the pins on a ESP32-WROOM-32.
So, if it is not running another thread maybe it is jumping out of the while loop.

If you put a print statement in the while loop and leave it at 2ms and then press buttonA do the print statements pause?

Doing `from m5stack import *` means everything comes in.  You need someone who knows M5 Stack.

**Reply by davefes** [reply]
At a REPL prompt:
`>>> import m5stack`
enter
`>>> dir(m5stack)`
enter

**Comment by sosi-deadeye** [comment, marked as answer]
I guess the following method adds an interrupt service routine to the GPIO:
```btnA.wasPressed(buttonA_wasPressed)```

This works even, if your main thread is blocked. This interrupts the main thread and the ISR is called and after this, the main thread continues the work.

But you can test it easily:

```python
from m5stack import *
from m5ui import *
from uiflow import *


setScreenColor(0x111111)
label0 = M5TextBox(13, 9, "label0", lcd.FONT_Default, 0xFFFFFF, rotate=0)


def buttonA_wasPressed():
    label0.setText('Hello M5')

btnA.wasPressed(buttonA_wasPressed)

# busy loop
while True:
    pass
```
If then a button press still changes the text, then it's an ISR.

**Comment by cristian-spiescu** [comment]
Thanks!

I'd say it certainly looks like an ISR.

I worked w/ ISRs in the past, on an ESP8266. I recall that code within an ISR is very very sensible. E.g. should be very quick; shouldn't modify the stack or something like that, etc. 1/ The doc was really explicit on this matter. 2/ I also got the taste of it. Was performing some string concatenations that would generate subtle corruptions of my app.

My hesitation came from the following fact: the WYSIWYG editor makes use of such handlers pretty nonchalantly. There are no warnings about what operations should or shouldn't be done in such handlers. So I make the assumptions that they cannot be ISRs. 

So the current hypothesis is that: the handlers are indeed ISRs. And probably that building apps a bit more heavier w/ the M5 stack WYSIWYG editor would probably do infringe the rules of ISR, generating strange runtime bugs.

**Reply by davefes** [reply]
I had read that doing:
`from m5stack import *`
was not a good way of accessing the few modules you might want to use.
